[quant] Add check for asan and qnnpack in the conv tests

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45688 [quant] slow path for the FBGEMM conv_transpose
* **#45913 [quant] Add check for asan and qnnpack in the conv tests**

Differential Revision: [D24142132](https://our.internmc.facebook.com/intern/diff/D24142132)

Commit:
[quant] Add check for asan and qnnpack in the conv tests

[ghstack-poisoned]

z-a-f commented on test/quantization/test_quantized_op.py:
```diff
@@ -3621,7 +3626,7 @@ def test_qconv1d(
         output_channels = output_channels_per_group * groups
         if torch.backends.quantized.engine == 'qnnpack':
-            use_channelwise = False
+            assume(not use_channelwise)
```
Whoops -- this is contamination

Commit:
Update on "[quant] Add check for asan and qnnpack in the conv tests"


Differential Revision: [D24142132](https://our.internmc.facebook.com/intern/diff/D24142132)

[ghstack-poisoned]

Commit:
Update on "[quant] Add check for asan and qnnpack in the conv tests"


Differential Revision: [D24142132](https://our.internmc.facebook.com/intern/diff/D24142132)

[ghstack-poisoned]

Commit:
Update on "[quant] Add check for asan and qnnpack in the conv tests"


Differential Revision: [D24142132](https://our.internmc.facebook.com/intern/diff/D24142132)

[ghstack-poisoned]

supriyar commented:
Can you elaborate a little on what the reason for disabling UBSAN for qnnpack tests is? Currently seems like we run the qconv tests using address sanitizer in CI.
@kimishpatel are you aware of any known issues with QNNPACK and UBSAN?

kimishpatel commented:
> Can you elaborate a little on what the reason for disabling UBSAN for qnnpack tests is? Currently seems like we run the qconv tests using address sanitizer in CI.
> @kimishpatel are you aware of any known issues with QNNPACK and UBSAN?

Zafar had pointed me to this failure at some point: https://app.circleci.com/pipelines/github/pytorch/pytorch/212765/workflows/46100497-5512-40d0-8f57-eb7f28d6ad6d/jobs/7401012. That complain does not seem legit though. The point is uint8 and it says uint32. I did not dig further.
But it seems disaling ubsan has been done before: https://github.com/pytorch/pytorch/blob/master/test/quantization/test_quantized_functional.py#L95

vkuzo commented:
> Can you elaborate a little on what the reason for disabling UBSAN for qnnpack tests is? Currently seems like we run the qconv tests using address sanitizer in CI.
> @kimishpatel are you aware of any known issues with QNNPACK and UBSAN?

+1, can we add the context of why we are disabling these tests to the PR summary as well?

kimishpatel commented:
> > Can you elaborate a little on what the reason for disabling UBSAN for qnnpack tests is? Currently seems like we run the qconv tests using address sanitizer in CI.
> > @kimishpatel are you aware of any known issues with QNNPACK and UBSAN?
> 
> +1, can we add the context of why we are disabling these tests to the PR summary as well?

So disabling UBSAN has been done earlier as well. I dont know the history there as to why it was disabled. For example in `test_quantized_functional.py` that I linked above.

As for the failure I linked above, which Zafar had reached out to me for, here is the context:
Failure comes from https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qnnpack/src/q8conv/4x4c2-sse2.c#L335.
Since pointer is being cast to uint_32 ubsan complains that memory address is not aligned to pointer type. It is perfectly possible that c1 will be unaligned to 32 bit type since 1) it is uint8 input pointer and 2) This arithmetic here https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qnnpack/src/q8conv/4x4c2-sse2.c#L321 can lead it to be unaligned to 32bit pointer even if the input int8 pointer is 32bit aligned.

Commit:
Update on "[quant] Add check for asan and qnnpack in the conv tests"


Differential Revision: [D24142132](https://our.internmc.facebook.com/intern/diff/D24142132)

[ghstack-poisoned]

Commit:
Update on "[quant] Add check for asan and qnnpack in the conv tests"


Differential Revision: [D24142132](https://our.internmc.facebook.com/intern/diff/D24142132)

[ghstack-poisoned]

Commit:
Update on "[quant] Add check for asan and qnnpack in the conv tests"


Differential Revision: [D24142132](https://our.internmc.facebook.com/intern/diff/D24142132)

[ghstack-poisoned]
